hexrd/xrdutil/nf-hedm/nf-HEDM_test.py

```diff
@@ -0,0 +1,1132 @@
+    """getter functions that handles the caching of the simulation"""
+    try:
+        image_stack = np.load(cache_file, mmap_mode='r', allow_pickle=False)
+    except Exception:
+        image_stack = simulate_diffractions(grain_params, experiment,
+                                            controller=controller)
+        np.save(cache_file, image_stack)
+
+    controller.handle_result('image_stack', image_stack)
+
+    return image_stack
+
+
+def simulate_diffractions(grain_params, experiment, controller):
+    """actual forward simulation of the diffraction"""
+
+    # use a packed array for the image_stack
+    array_dims = (experiment.nframes,
+                  experiment.ncols,
+                  ((experiment.nrows - 1)//8) + 1)
+    image_stack = np.zeros(array_dims, dtype=np.uint8)
+
+    count = len(grain_params)
+    subprocess = 'simulate diffractions'
+
+    _project = xrdutil._project_on_detector_plane
+    rD = experiment.rMat_d
+    chi = experiment.chi
+    tD = experiment.tVec_d
+    tS = experiment.tVec_s
+    distortion = experiment.distortion
+
+    eta_range = [(-np.pi, np.pi), ]
+    ome_range = experiment.ome_range
+    ome_period = (-np.pi, np.pi)
+
+    full_hkls = xrdutil._fetch_hkls_from_planedata(experiment.plane_data)
+    bMat = experiment.plane_data.latVecOps['B']
+    wlen = experiment.plane_data.wavelength
+
+    controller.start(subprocess, count)
+    for i in range(count):
+        rC = xfcapi.makeRotMatOfExpMap(grain_params[i][0:3])
+        tC = np.ascontiguousarray(grain_params[i][3:6])
+        vInv_s = np.ascontiguousarray(grain_params[i][6:12])
+        ang_list = np.vstack(xfcapi.oscillAnglesOfHKLs(full_hkls[:, 1:], chi,
+                                                       rC, bMat, wlen,
+                                                       vInv=vInv_s))
+        # hkls not needed here
+        all_angs, _ = xrdutil._filter_hkls_eta_ome(full_hkls, ang_list,
+                                                   eta_range, ome_range)
+        all_angs[:, 2] = xfcapi.mapAngle(all_angs[:, 2], ome_period)
+
+        proj_pts =  _project(all_angs, rD, rC, chi, tD,
+                             tC, tS, distortion)
+        det_xy = proj_pts[0]
+        _write_pixels(det_xy, all_angs[:, 2], image_stack, experiment.base,
+                      experiment.inv_deltas, experiment.clip_vals)
+
+        controller.update(i + 1)
+
+    controller.finish(subprocess)
+    return image_stack
+
+
+# This part is critical for the performance of simulate diffractions. It
+# basically "renders" the "pixels". It takes the coordinates, quantizes to an
+# image coordinate and writes to the appropriate image in the stack. Note
+# that it also performs clipping based on inv_deltas and clip_vals.
+#
+# Note: This could be easily modified so that instead of using an array of
+#       booleans, an array of uint8 could be used so the image is stored
+#       with a bit per pixel.
+
+@numba.njit
```

Comment:
For these numba functions, we might be able to get a little bit of a speed-up from using `cache=True`, like [here](https://github.com/HEXRD/hexrd/blob/0a0886e84bee8094e0a544e181fbaae25a928f76/hexrd/matrixutil.py#L855).

The issues with caching on Windows in hexrd should all be resolved. We can benchmark and test it though.

Reply:
good point